nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,209 @@
+              Whether to disable user registration to the server
+              without the shared secret.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for dendrite, see:
+        <link xlink:href="https://github.com/matrix-org/dendrite/blob/master/dendrite-config.yaml"/>
+        for available options with which to populate settings.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.matrix-dendrite.tlsCert = lib.mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.cert";
+    services.matrix-dendrite.tlsKey = lib.mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.key";
+
+    systemd.services.matrix-dendrite = {
+      description = "Dendrite Matrix homeserver";
+      after = [
+        "network.target"
+      ];
+      wantedBy = [ "multi-user.target" ];
+      preStart    = ''
+        ${pkgs.envsubst}/bin/envsubst \
+          -i ${configurationYaml} \
+          -o /run/matrix-dendrite/dendrite.yaml
+      '';
+      serviceConfig = {
+        Type = "simple";
+        DynamicUser = true;
+        StateDirectory = "matrix-dendrite";
+        WorkingDirectory = workingDir;
+        ExecReload = "kill -HUP $MAINPID";
+        EnvironmentFile = lib.mkIf (cfg.environmentFile != null) cfg.environmentFile;
+        ExecStart = lib.strings.concatStringsSep " " ([
+          "${pkgs.matrix-dendrite}/bin/dendrite-monolith-server"
+          "--config /run/matrix-dendrite/dendrite.yaml"
+        ] ++ lib.optionals (cfg.httpPort != null) [
+          "--http-bind-address :${builtins.toString cfg.httpPort}"
+        ] ++ lib.optionals (cfg.httpsPort != null) [
+          "--https-bind-address :${builtins.toString cfg.httpsPort}"
+        ] ++ lib.optionals (cfg.tlsCert != null && cfg.tlsKey != null) [
+          "--tls-cert ${cfg.tlsCert}"
+          "--tls-key ${cfg.tlsKey}"
+        ]);
+        Restart = "on-failure";
+        RuntimeDirectory = "matrix-dendrite";
+        RuntimeDirectoryMode = "0700";
+      };
+    };
+    systemd.services.matrix-dendrite-key-generator =
+      lib.mkIf (cfg.generatePrivateKey || cfg.generateTls) {
+        description = "Dendrite Matrix homeserver - Key Generator";
+        requiredBy = [ "matrix-dendrite.service" ];
+        before = [ "matrix-dendrite.service" ];
+        path = with pkgs; [ matrix-dendrite ];
+        serviceConfig = {
+          Type = "oneshot";
+          DynamicUser = true;
+          StateDirectory = "matrix-dendrite";
+          WorkingDirectory = workingDir;
+          RemainAfterExit = true;
+          ExecStart = lib.strings.concatStringsSep " " ([
+            "${pkgs.matrix-dendrite}/bin/generate-keys"
+          ] ++ lib.optionals cfg.generatePrivateKey [
+            "--private-key ${cfg.settings.global.private_key}"
+          ] ++ lib.optionals cfg.generateTls [
+            "--tls-cert ${cfg.tlsCert}"
+            "--tls-key ${cfg.tlsKey}"
+          ]);
```

Comment:
Won't this overwrite the key after a reboot? Maybe instead we should only run this if the key does not exist

```
unitConfig.ConditionPathExists = "!${cfg.settings.global.private_key}";
```

In that case I would also remove `RemainAfterExit = true` so this check is performed each time dendrite starts.

What do you think?

Reply:
Not sure what to do about this :/

Reply:
Maybe turn it into a shell script that generates each of those files separately when missing. Or you could just simplify the whole thing and remove TLS generation which I think was discussed previously in the PR. Self-signed certs are not very useful anyway and most people will use LetsEncrypt. In that case the `ConditionPathExists` trick would work.

Reply:
I think anyone deploying this is going to be bringing their own certs anyway? I'd be keen to simplify

Reply:
Is it worth having the one-shot service for the private-key at that point?